packages/contracts-bedrock/src/Safe/LivenessGuard.sol

```diff
@@ -0,0 +1,135 @@
+        }
+
+        // This call will reenter to the Safe which is calling it. This is OK because it is only reading the
+        // nonce, and using the getTransactionHash() method.
+        bytes32 txHash = Safe(payable(msg.sender)).getTransactionHash(
+            // Transaction info
+            to,
+            value,
+            data,
+            operation,
+            safeTxGas,
+            // Payment info
+            baseGas,
+            gasPrice,
+            gasToken,
+            refundReceiver,
+            // Signature info
+            Safe(payable(msg.sender)).nonce() - 1
+        );
+        address[] memory signers = _getNSigners(txHash, signatures);
+        for (uint256 i = 0; i < signers.length; i++) {
+            lastSigned[signers[i]] = block.timestamp;
+        }
+        emit SignersRecorded(txHash, signers);
+    }
+
+    /// @notice Exctract the signers from a set of signatures.
+    function _getNSigners(bytes32 dataHash, bytes memory signatures) internal pure returns (address[] memory _owners) {
+        uint256 numSignatures = signatures.length / 65;
+        _owners = new address[](numSignatures);
+
+        /// The following code is extracted from the Safe.checkNSignatures() method. It removes the signature
+        /// validation code, and keeps only the parsing code necessary to extract the owner addresses from the
+        /// signatures. We do not double check if the owner derived from a signature is valid. As this is handled
+        /// in the final require statement of Safe.checkNSignatures().
+        address currentOwner;
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+        uint256 i;
+        for (i = 0; i < numSignatures; i++) {
+            (v, r, s) = signatureSplit(signatures, i);
+            if (v == 0) {
+                // If v is 0 then it is a contract signature
+                // When handling contract signatures the address of the contract is encoded into r
+                currentOwner = address(uint160(uint256(r)));
+            } else if (v == 1) {
+                // If v is 1 then it is an approved hash
+                // When handling approved hashes the address of the approver is encoded into r
+                currentOwner = address(uint160(uint256(r)));
+            } else if (v > 30) {
+                // If v > 30 then default va (27,28) has been adjusted for eth_sign flow
+                // To support eth_sign and similar we adjust v and hash the messageHash with the Ethereum message prefix
+                // before applying ecrecover
+                currentOwner =
+                    ecrecover(keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", dataHash)), v - 4, r, s);
+            } else {
+                // Default is the ecrecover flow with the provided data hash
+                // Use ecrecover with the messageHash for EOA signatures
+                currentOwner = ecrecover(dataHash, v, r, s);
+            }
```

Comment:
This should revert if `currentOwner == address(0)` at this point, allowing this seems like it would only enable some sort of bug

Reply:
Good point. I agree that we don't want a non-zero timestamp indicating that the zero-address has signed anything!

For context, that `ecrecover` is taken from the Safe contract. It's safe there because `currentOwner == address(0)` would cause an error in the last check which enforces the ordering of the signatures (`require(currentOwner > lastOwner...`. 

The check you suggest is most likely safe to include, but it is somewhat risky to add a `revert` in this codepath, as we're only trying to record signatures, not prevent a transaction from executing. IMO, a more targeted mitigation would be to detect when a signer is returned as `address(0)`, and skip writing to the `lastSigned` mapping. 

https://github.com/safe-global/safe-contracts/blob/e870f514ad34cd9654c72174d6d4a839e3c6639f/contracts/Safe.sol#L326-L333

```
     } else {
          // Default is the ecrecover flow with the provided data hash
          // Use ecrecover with the messageHash for EOA signatures
          currentOwner = ecrecover(dataHash, v, r, s);
      }
      require(currentOwner > lastOwner && owners[currentOwner] != address(0) && currentOwner != SENTINEL_OWNERS, "GS026");
      lastOwner = currentOwner;
  }
```

I suppose it's OK there because of the `currentOwner > lastOwner` check.

Reply:
In discussion we agreed it was important to avoid any situation which could lead to a revert in the guard. Resolving.